generator/migration.go

```diff
@@ -0,0 +1,670 @@
+	schema := struct {
+		Tables []*TableSchema
+	}{s.Tables}
+	return json.MarshalIndent(schema, "", "  ")
+}
+
+// Table finds a table with the given name.
+func (s *ModelSchema) Table(name string) *TableSchema {
+	for _, t := range s.Tables {
+		if t.Name == name {
+			return t
+		}
+	}
+	return nil
+}
+
+// TableSchema represents the SQL schema of a table.
+type TableSchema struct {
+	// Name is the table name.
+	Name string
+	// Columns are the schemas of the columns in the table.
+	Columns []*ColumnSchema
+}
+
+func (s *TableSchema) String() string {
+	var buf bytes.Buffer
+	buf.WriteString(fmt.Sprintf("CREATE TABLE %s (\n", s.Name))
+	for i, c := range s.Columns {
+		buf.WriteRune('\t')
+		buf.WriteString(c.String())
+		if i < len(s.Columns)-1 {
+			buf.WriteString(",\n")
+		} else {
+			buf.WriteRune('\n')
+		}
+	}
+	buf.WriteString(");\n\n")
+	return buf.String()
+}
+
+// Columns returns the schema of the column with the given name.
+func (s *TableSchema) Column(name string) *ColumnSchema {
+	for _, c := range s.Columns {
+		if c.Name == name {
+			return c
+		}
+	}
+	return nil
+}
+
+func (s *TableSchema) Equals(s2 *TableSchema) bool {
+	if s.Name != s2.Name || len(s.Columns) != len(s2.Columns) {
+		return false
+	}
+
+	for i, c := range s.Columns {
+		if !c.Equals(s2.Columns[i]) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// ColumnSchema represents the schema of a column.
+type ColumnSchema struct {
+	// Name of the column.
+	Name string
+	// Type of the column.
+	Type ColumnType
+	// PrimaryKey reports whether the column is a primary key.
+	PrimaryKey bool
+	// Reference is an optional reference to another table column.
+	// If it's not nil, it means this column has a foreign key.
+	Reference *Reference
+	// NotNull reports whether the column is not nullable.
+	NotNull bool
+}
+
+func (s *ColumnSchema) Equals(s2 *ColumnSchema) bool {
+	return s.Name == s2.Name &&
+		s.Type == s2.Type &&
+		s.PrimaryKey == s2.PrimaryKey &&
+		s.NotNull == s2.NotNull &&
+		s.Reference.Equals(s2.Reference)
+}
+
+func (s *ColumnSchema) String() string {
+	var buf bytes.Buffer
+	buf.WriteString(s.Name)
+	buf.WriteRune(' ')
+	buf.WriteString(string(s.Type))
+
+	if s.NotNull {
+		buf.WriteString(" NOT NULL")
+	}
+
+	if s.PrimaryKey {
+		buf.WriteString(" PRIMARY KEY")
+	}
+
+	if s.Reference != nil {
+		buf.WriteString(" REFERENCES ")
+		buf.WriteString(s.Reference.String())
+	}
+
+	return buf.String()
+}
+
+// ColumnType represents the SQL column type.
+type ColumnType string
+
+const (
+	SmallIntColumn    ColumnType = "smallint"
+	IntegerColumn     ColumnType = "integer"
+	BigIntColumn      ColumnType = "bigint"
+	RealColumn        ColumnType = "real"
+	DoubleColumn      ColumnType = "double"
+	SmallSerialColumn ColumnType = "smallserial"
+	SerialColumn      ColumnType = "serial"
+	BigSerialColumn   ColumnType = "bigserial"
+	TimestamptzColumn ColumnType = "timestamptz"
+	TextColumn        ColumnType = "text"
```

Comment:
For short strings could be better use a `varchar(n)` instead of `text`, How will we able to choose this?